cmd/driver-registrar/main.go

```diff
@@ -50,14 +56,122 @@ const (
-	showVersion       = flag.Bool("version", false, "Show version.")
-
-	version = "unknown"
+	kubeconfig              = flag.String("kubeconfig", "", "Absolute path to the kubeconfig file. Required only when running out of cluster.")
+	connectionTimeout       = flag.Duration("connection-timeout", 1*time.Minute, "Timeout for waiting for CSI driver socket.")
+	csiAddress              = flag.String("csi-address", "/run/csi/socket", "Address of the CSI driver socket.")
+	kubeletRegistrationPath = flag.String("kubelet-registration-path", "/var/lib/kubelet/plugins/csi-hostpath/csi.sock", "Full path to CSI driver's socket.")
+	showVersion             = flag.Bool("version", false, "Show version.")
+	version                 = "unknown"
+	// List of supported versions
+	supportedVersions = []string{"0.2.0", "0.3.0"}
 )
 
+// registrationServer is a sample plugin to work with plugin watcher
+type registrationServer struct {
+	driverName string
+	grpcServer *grpc.Server
+	wg         sync.WaitGroup
+	endpoint   string
+	version    []string
+}
+
+// NewregistrationServer returns an initialized registrationServer instance
+func newRegistrationServer() *registrationServer {
+	return &registrationServer{}
+}
+
+// GetInfo is the RPC invoked by plugin watcher
+func (e *registrationServer) GetInfo(ctx context.Context, req *registerapi.InfoRequest) (*registerapi.PluginInfo, error) {
+
+	glog.Infof("><SB> Received GetInfo call: %+v", req)
+
+	return &registerapi.PluginInfo{
+		Type:              registerapi.CSIPlugin,
+		Name:              e.driverName,
+		Endpoint:          e.endpoint,
+		SupportedVersions: e.version,
+	}, nil
+}
+
+func (e *registrationServer) NotifyRegistrationStatus(ctx context.Context, status *registerapi.RegistrationStatus) (*registerapi.RegistrationStatusResponse, error) {
+
+	glog.Infof("><SB> Received NotifyRegistrationStatus call: %+v", status)
+
+	if !status.PluginRegistered {
+		glog.Errorf("Registration process failed with error: %+v, restarting registration container.", status.Error)
+		os.Exit(1)
+	}
+
+	return &registerapi.RegistrationStatusResponse{}, nil
+}
+
+// Serve starts Registration grpc server
+func (e *registrationServer) Serve(socketPath string) error {
+	// Remove any socket, stale or not, but fall through for other files
+	fi, err := os.Stat(socketPath)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+		os.Remove(socketPath)
+	}
+	// Default to only user accessible socket, caller can open up later if desired
+	oldmask := unix.Umask(0077)
+
+	glog.Infof("Starting Registration Server at: %s\n", socketPath)
+	lis, err := net.Listen("unix", socketPath)
+	if err != nil {
+		return err
+	}
+	unix.Umask(oldmask)
+	glog.Infof("Registration Server started at: %s\n", socketPath)
+	e.grpcServer = grpc.NewServer()
+	// Registers kubelet plugin watcher api.
+	registerapi.RegisterRegistrationServer(e.grpcServer, e)
+
+	// Starts service
+	e.wg.Add(1)
```

Comment:
Why do we need wait groups here at all? Would a simple channel suffice since we only ever add to it once and just `Done()` on it after the goroutine below finishes.